<div width=50% style="display: block; margin: auto">
    <img src="figures/ucl-logo.svg" width=100%>
</div>


### [UCL-ELEC0136 Data Acquisition and Processing Systems 2024]()
University College London
# Lab 4: Advanced Data Storage - Vector Databases and LLMs


<hr width=70% style="float: left">

**IMPORTANT:** The content of this Notebook will not be evaluated in the final exam. The goal is to provide you with practical experience with the very trendy subject that LLMs are, for you to know how to use them, and provide you with enough to start your own LLM projects. 

This lab also serves as a good illustration of what this module is about: how data acquisition, storage, and processing all come together to create inteligent AI-driven applications.

### Objectives
* Perform CRUD operations on a Pinecone Vector Database.
* Use Langchain to make queries to API accessed pre-trained LLM models (from Hugging Face and OpenAI).
* Compare the performances of two pre-trained LLMs.
* Use a Pinecone vector Database to perform Retrieval Augmentation of a pre-trained LLM, allowing it to both expend his knowledge base, and cite sources.

### Outline

This notebook has 3 parts:

0. [Setting up](#0.-Setting-up)
1. [CRUD operations on a vector database](#1-crud-operations-on-a-pinecone-vector-database)
2. [Intro to LLMs and LangChain](#2.-Intro-to-LLMs-and-Langchain)
3. [Retrieval Augmentation of a LLM using a vector database](#3.-Retrieval-Augmentation-of-a-LLM-using-a-vector-database)

<hr width=70% style="float: left">

# 0. Setting up

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- The assignment repository contains a `requirements.txt` file, make sure to install all the librairies with the correct versions listed in this file in your daps conda environment.

</div>

## 0.1 Create a Pinecone account and connect to your free-tier online vector database

[Pinecone](https://www.pinecone.io) is a vector database service that helps developers build and deploy applications with high-performance similarity search and recommendation capabilities. It enables efficient storage and retrieval of vector data, making it easier to create personalized experiences and content recommendations in various applications, such as stable diffusion, LLMs chatbox, and many other AI applications.

In this Notebook, we will use Pinecone's free tier to create and connect to a vector database, perform CRUD operations, and then use that database to power a LLM application.




<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Follow the instructions in `pinecone_tutorial.pdf` to create a Pinecone account and get an API key for the free tier vector database. MAKE SURE TO KEEP A COPY OF YOUR API KEY.
- Run the cell bellow to connect to your online vector database.

</div>

In [7]:
# Run this cell (it may take a few seconds)
import pinecone

In [15]:
###########################
# Task: 
#   change PINECONE_API_KEY with your pinecone API key and run the cell
#
###########################

PINECONE_API_KEY = "210bffc8-1c2f-43b8-a7ee-5e513bdbb2c0" #<--- TODO: your API key here 

pinecone.init(api_key=PINECONE_API_KEY, environment="gcp-starter")

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`list_indexes`](https://docs.pinecone.io/reference/list_indexes) function to return the list of Pinecone indexes you have on your database (it should be empty).
- If it isn't empty, use the [`delete_index`](https://docs.pinecone.io/reference/delete_index) function to delete any indexes you may have on your database.

</div>

In [14]:
###########################
# Task: 
#   Check that your pinecone database does not contain any indexes, and delete them if there are any.
#
###########################

# TODO : your code bellow
pinecone.list_indexes()

['langchain-retrieval-augmentation-fast']

## 0.2 Create a HuggingFace account and generate a free API key

**Note:** you do not need this step to do part. [# 1. CRUD opperations on a Vector Database](#1.-CRUD-opperations-on-a-Vector-Database).


[Hugging Face 🤗](https://huggingface.co) is a company and open-source platform that specializes in natural language processing (NLP) and provides tools, libraries, and pre-trained models for building and deploying NLP applications. Their most well-known product is the Transformers library, which offers access to a wide range of pre-trained NLP models, making it easier for developers to work with text-based tasks such as language translation, sentiment analysis, and more.

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Follow the instructions in `huggingface_tutorial.pdf` to create a HuggingFace account and get an API key for the free tier vector database. MAKE SURE TO KEEP A COPY OF YOUR API KEY.
- Run the cell bellow to set an environment variable to your API key. `langchain.HuggingFaceHub` will use this environment variable when sending a request to the Hugging Face server to authentificate the connection.

</div>

In [13]:
import os

###########################
# Task: 
#   change HUGGING_FACE_API_KEY with your Hugging Face API key and run the cell to set the environment variable HUGGINGFACEHUB_API_TOKEN
#
###########################

HUGGING_FACE_API_KEY = "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" #<--- TODO: your Hugging Face API key here 

os.environ['HUGGINGFACEHUB_API_TOKEN'] = HUGGING_FACE_API_KEY

## 0.3 Create an OpenAI account and generate a free API key

**Note:** you do not need this step to do part. [1. CRUD opperations on a Vector Database](#1.-crud-operations-on-a-pinecone-vector-database) and part. [2.1 Hugging Face LLM](#211-initializing-the-llm).

[OpenAI](https://openai.com) is an artificial intelligence (AI) research laboratory consisting of the for-profit OpenAI LP and its non-profit parent company, OpenAI Inc.

OpenAI provides an [API](https://platform.openai.com/docs/overview) that allows developers to access and integrate the capabilities of OpenAI's language models into their own applications, products, or services. The OpenAI API is based on models like GPT and allows developers to make use of powerful natural language processing (NLP) functionalities.

**Unlike Hugging Face, OpenAI is not open source, and the free tier of their API is only available for 3 months after the creation of a new account, and has a lot of restriction (for exemple, you are limited to 3 requests per minutes for each model). However, their models are state of the art and very powerful, which is why we will use them in this lab.**

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Follow the instructions in `openai_tutorial.pdf` to create a HuggingFace account and get an API key for the free tier vector database. MAKE SURE TO KEEP A COPY OF YOUR API KEY.
- Run the cell bellow to set an environment variable to your API key. 

</div>

In [12]:
###########################
# Task: 
#   change OPENAI_API_KEY with your OpenAI API key and run the cell to set the environment variable HUGGINGFACEHUB_API_TOKEN
#
###########################


OPENAI_API_KEY = "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" #<--- TODO: your OpenAI API key here 

os.environ['OPENAI_API_KEY'] = OPENAI_API_KEY

# 1. CRUD operations on a Pinecone Vector Database

*Source: https://docs.pinecone.io/docs/quickstart*


In this part, we will familiarize ourselve with basic operations on a Pinecode Vector Database.

<div class="alert alert-block alert-warning">
    <b>👩‍💻👨‍💻 Optional action</b>

- Very short video intro on Vector Databases: [Vector databases are so hot right now. What are they?](https://www.youtube.com/watch?v=klTvEwg3oJ4)
- Short read: [Everything you need to know about Pinecone – A Vector Database](https://www.packtpub.com/article-hub/everything-you-need-to-know-about-pinecone-a-vector-database).
- In depth read: [What is a Vector Database & How Does it Work? Use Cases + Examples](https://www.pinecone.io/learn/vector-database/).

</div>

<div class="alert alert-heading alert-danger" style="background-color: white; border: 2px solid; border-radius: 5px; color: #000; border-color:#AAA; padding: 10px">
    <b>💎 Tip</b>

Once your connection to your Pinecone database is established (which should have been done in part. 0.1), you do not need to use `pinecone.init` anymore and can directly use the API requests.

</div>

### 1.1 CRUD operations on a Vector Database - Create

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`create_index`](https://docs.pinecone.io/reference/create_index) function to create an index, name it `quickstart`, set the dimension to 8, and use the metric `euclidean`.
- Check that your database contains an index called `quickstart`.
- Use the [`describe_index`](https://docs.pinecone.io/reference/describe_index) function to get informations about the index `quickstart`.

</div>

In [4]:
###########################
# Task: 
#   Create an index called quickstart, check that it has been added to your Pinecone Vector DB, and get information about that index
#
###########################

# TODO : your code bellow
#DAPSweek4 = _get_api_instance()
pinecone.create_index(name='quickstart',dimension = 8, metric='euclidean')

pinecone.list_indexes()

pinecone.describe_index(name='quickstart')

NameError: name 'pinecone' is not defined

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`Index`](https://docs.pinecone.io/docs/python-client#index) class to contruct an `Index` object from the index `quickstart`. 
- Use the [`upsert`](https://docs.pinecone.io/docs/python-client#indexupsert) method to push the 5 vectors in the cell bellow to the index `quickstart`.
- Use the [`describe_index_stats`](https://docs.pinecone.io/docs/python-client#indexdescribe_index_stats) method to get statistics about the index's contents.

</div>


<div class="alert alert-heading alert-danger" style="background-color: white; border: 2px solid; border-radius: 5px; color: #000; border-color:#AAA; padding: 10px">
    <b>💎 Tip</b>

Wait for a few seconds after you use `upsert` before querying the index with `describe_index_stats` as the data needs a bit of time to be saved.

</div>

In [ ]:
###########################
# Task: 
#   Construct an Index object and use it to push the 5 vectors in data to your index, and get statistics about the index
#
###########################
index_me = pinecone.Index("quickstart")


### 1.2 CRUD operations on a Vector Database - Read

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`query`](https://docs.pinecone.io/docs/python-client#indexquery) method to search for the 3 closest vectors to the `target_vector` in the index `quickstart`.

</div>

In [ ]:
###########################
# Task: 
#   find the 3 nearest vectors to the target_vector
#
###########################
target_vector = [0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3]

index_me.query(target_vector,top_k= 3)
# TODO : your code bellow


{'matches': [{'id': 'C', 'score': 0.0, 'values': []},
             {'id': 'D', 'score': 0.0799999237, 'values': []},
             {'id': 'B', 'score': 0.0800000429, 'values': []}],
 'namespace': ''}

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`query`](https://docs.pinecone.io/docs/python-client#indexquery) method to search for the 3 closest vectors to the `target_vector` the index `quickstart` that are dramas.

</div>

In [ ]:
###########################
# Task: 
#   find the 3 nearest vectors to the target_vector that are dramas
#
###########################

index_me.query(target_vector,top_k= 3,filter={
        'genre': {'$in': ['drama']}
        })
# TODO : your code bellow


{'matches': [{'id': 'D', 'score': 0.0799999237, 'values': []},
             {'id': 'E', 'score': 0.319999695, 'values': []}],
 'namespace': ''}

### 1.3 CRUD operations on a Vector Database - Update

We want to change the values associated to vectors A and D.

In [ ]:
# Run this cell
print(index_me.query(
  id = "A",
  top_k = 1,
  include_values = True,
    include_metadata = True
))

print(index_me.query(
  id = "D",
  top_k = 1,
  include_values = True,
  include_metadata = True
))

{'matches': [{'id': 'A',
              'metadata': {'genre': 'comedy', 'year': 2020.0},
              'score': 0.0,
              'values': [0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]}],
 'namespace': ''}
{'matches': [{'id': 'D',
              'metadata': {'genre': 'drama'},
              'score': 0.0,
              'values': [0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4]}],
 'namespace': ''}


<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`update`](https://docs.pinecone.io/docs/python-client#indexupdate) method to do the following updates:
    - Replace all the values of `A` by `0.6`, change the genre to `action-comedy`.
    - Replace all the values of `D` by `-0.4`, add a `year` field in the metadata and set it to `2019`.

</div>

<div class="alert alert-heading alert-danger" style="background-color: white; border: 2px solid; border-radius: 5px; color: #000; border-color:#AAA; padding: 10px">
    <b>💎 Tip</b>

Wait for a few seconds after you use `update` before querying the index as the data needs a bit of time to be saved.

</div>

In [ ]:
update_size = index_me.describe_index_stats().dimension
update_size

8

In [ ]:
###########################
# Task: 
#   Update vectors A and D as indicated above.
#
###########################


# TODO : your code bellow
#A_all = index_me.fetch(ids=['A'])
#A_all['values']


In [ ]:
import numpy as np
A_update = 0.6*np.ones(update_size)
D_update = -0.4*np.ones(update_size)
print(A_update)

[0.6 0.6 0.6 0.6 0.6 0.6 0.6 0.6]


In [ ]:
list(A_update)

[0.6, 0.6, 0.6, 0.6, 0.6, 0.6, 0.6, 0.6]

In [ ]:
import time
update_response = index_me.update(
    id='A',
    values = list(A_update),
    set_metadata={'genre': 'action-comedy'}
)

update_response = index_me.update(
    id='D',
    values = list(D_update),
    set_metadata={'year': 2019}
)

time.sleep(5)

NameError: name 'index_me' is not defined

In [ ]:
# Run this cell
print(index_me.query(
  id = "A",
  top_k = 1,
  include_values = True,
  include_metadata = True
))

print(index_me.query(
  id = "D",
  top_k = 1,
  include_values = True,
  include_metadata = True
))

{'matches': [{'id': 'A',
              'metadata': {'genre': 'action-comedy', 'year': 2020.0},
              'score': 0.0,
              'values': [0.6, 0.6, 0.6, 0.6, 0.6, 0.6, 0.6, 0.6]}],
 'namespace': ''}
{'matches': [{'id': 'D',
              'metadata': {'genre': 'drama', 'year': 2019.0},
              'score': 0.0,
              'values': [-0.4, -0.4, -0.4, -0.4, -0.4, -0.4, -0.4, -0.4]}],
 'namespace': ''}


### 1.4 CRUD operations on a Vector Database - Delete

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`delete`](https://docs.pinecone.io/docs/python-client#indexdelete) method to delete the vectors `B` and `C`.
- Check with `describe_index_stats` that your index now contains 3 vectors

</div>

<div class="alert alert-heading alert-danger" style="background-color: white; border: 2px solid; border-radius: 5px; color: #000; border-color:#AAA; padding: 10px">
    <b>💎 Tip</b>

Wait for a few seconds after you use `upsert` before querying the index with `describe_index_stats` as the data needs a bit of time to be saved.

</div>

In [ ]:
###########################
# Task: 
#   Delete vectors B and C from the index.
#
###########################


# TODO : your code bellow
delete_response = index_me.delete(ids= ['B','D'])

time.sleep(5)

In [ ]:
index_me.describe_index_stats()

{'dimension': 8,
 'index_fullness': 3e-05,
 'namespaces': {'': {'vector_count': 3}},
 'total_vector_count': 3}

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>
    
- Use the [`delete_index`](https://docs.pinecone.io/reference/delete_index) function to delete the index `quickstart`.

</div>

In [42]:
###########################
# Task: 
#   delete the index quickstart from your Pinecode Vector Database
#
###########################

# TODO : your code bellow
pinecone.delete_index('quickstart')


In [43]:

pinecone.list_indexes()

[]

# 2. Intro to LLMs and LangChain

**Context:**

* **Large language models (LLMs)** are a class of artificial intelligence models that have been trained on vast amounts of text data to understand and generate human-like text. These models are based on deep learning techniques, such as neural networks, and have many parameters, often numbering in the hundreds of millions or even billions. Some well-known examples include OpenAI's GPT-3 and GPT-4, and Google's BERT and T5 models.

* " [**LangChain**](https://python.langchain.com/docs/get_started/introduction) is an open source framework that lets software developers working with artificial intelligence (AI) and its machine learning subset combine large language models with other external components to develop LLM-powered applications. The goal of LangChain is to link powerful LLMs, such as OpenAI's GPT-3.5 and GPT-4, to an array of external data sources to create and reap the benefits of natural language processing (NLP) applications. " - [Source](https://www.techtarget.com/searchenterpriseai/definition/LangChain#:~:text=LangChain%20is%20an%20open%20source,to%20develop%20LLM%2Dpowered%20applications.)



**Although we are doing these operations in a Jupyter Notebook, the exact same code can be used to program server-hosted web applications that could perform real-world tasks.**


<div class="alert alert-block alert-warning">
    <b>👩‍💻👨‍💻 Optional action</b>

- In depth read: [LangChain AI Handbook](https://www.pinecone.io/learn/series/langchain/).

</div>


In this part, we will use LangChain to deploy two pre-trained LLMs, one from Hugging Face Hub, and one from OpenAI's API.

*Source: [https://www.pinecone.io/learn/series/langchain/langchain-intro/](https://www.pinecone.io/learn/series/langchain/langchain-intro/)*

## 2.1 Hugging Face LLM

Let's use a pre-trained Google language model, [flan-t5-xxl](https://huggingface.co/google/flan-t5-xxl), hosted on the Hugging Face Hub, that we will access for free through Hugging Face's API. More specifically, we will use the HuggingFaceHub module of the langchain library, which will query Hugging Face's API for us.

### 2.1.1 Initializing the LLM

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the cell below to initialize the connection to the LLM.

</div>

In [35]:
# Run this cell

from langchain import HuggingFaceHub

# initialize Hub LLM
hub_llm = HuggingFaceHub(
        repo_id='google/flan-t5-xxl',
    model_kwargs={'temperature':1e-2} #Best temperature found: 1e-2
)

c:\Anaconda\envs\daps\Lib\site-packages\huggingface_hub\utils\_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '0.19.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


### 2.1.2 Use LangChain to ask a question to the LLM

The first thing we need to do to query a LLM is to create a **prompt template**. A prompt template contains instructions to generate a prompt in a reproductible way. It contains a text string ("the template"), that can take in a set of parameters from the end user and generates a prompt (the input variables).

For example: 

* We want a LLM model to tell what language a sentence is written in, then an appropriate prompt template would be: 
    * `"What language is the sentense "{sentence}" written in?"`
    * Here, `sentence` is the only input variable. 

<br/>

* We want a LLM model to generate jokes about a topic, while specifying what type of jokes, then an appropriate prompt template would be: 
    * `"Make a {type} of joke about {subject}""`
    * Here, `type` and `subject` are the input variables.

<br/>


**Note:** `google/flan-t5-xxl` is a small LLM, best suited to give short answers. Although it is enough for us to explore LLMs today, for a real application bigger models would be better suited.

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Using [langchain.PromptTemplate](https://api.python.langchain.com/en/latest/prompts/langchain.prompts.prompt.PromptTemplate.html), create a prompt template with an input variable called `name` that asks the language model to give the date of birth of a historical figure. 

</div>

In [33]:
###########################
# Task: 
#   Create a PromptTemplate with an input variable called `name` that asks the language model to give the date of birth of a historical figure.
#
###########################

from langchain import PromptTemplate

# TODO : your code bellow

# Instantiation using from_template (recommended)
prompt_me = PromptTemplate.from_template("Please find the Date of Birth of {name}")
#prompt_me.format(name="me")

# Instantiation using initializer
#prompt = PromptTemplate(input_variables=["foo"], template="Say {foo}")



<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Using langchain.LLMChain, create a chain to run the prompt by our LLM.

</div>


<div class="alert alert-heading alert-danger" style="background-color: white; border: 2px solid; border-radius: 5px; color: #000; border-color:#AAA; padding: 10px">
    <b>💎 Tip</b>

An LLMChain consists of a PromptTemplate and a language model (either an LLM or chat model). It formats the prompt template using the input key values provided, passes the formatted string to LLM and returns the LLM output.

**Example:** 
llm_chain = LLMChain(
    prompt=your_PromptTemplate_template,
    llm=your_llm
)
</div>


In [36]:
###########################
# Task: 
#   Create a LLMchain with your prompt and your hub_llm
#
###########################

from langchain import LLMChain

# TODO : your code bellow
llm_Ch = LLMChain(prompt=prompt_me,llm = hub_llm)


<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Use `LLMChain.run()` or `LLMChain.predict()` to ask for the date of birth of Napoleon.


</div>

In [47]:
###########################
# Task: 
#   Asks the language model to give the date of birth of a historical figure.
#
###########################

# TODO : your code bellow
llm_Ch.run('Napoleaon')


'1769-01-06'

**SPOILER:** Well, that's not amazing. The LLM understood we wanted a date, and we even got something close to Napoleon's real date of birth, but the results is wrong. This is because this LLM is small and not adapted to this task.

## 2.2 OpenAI LLM

In this part, we will task a more powerfull openAI LLM called [`text-davinci-003`](https://platform.openai.com/docs/models/gpt-3), a variant of GPT3,to perform the same task we instructed the Hugging Face hosted model, and see if we get better results.

### 2.2.1 Initializing the LLM

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the cell bellow to innitialize the connection to the LLM.

</div>

In [30]:
# Run this cell

from langchain.llms import OpenAI

davinci = OpenAI(model_name='text-davinci-003', openai_api_key =  OPENAI_API_KEY)

### 2.2.2 Use LangChain to ask a question to the LLM

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Using langchain.LLMChain, create a chain to run the same prompt as in part. 2.1.2 by our OpenAI LLM.

</div>

In [ ]:
###########################
# Task: 
#   Create a LLMchain with your prompt and your hub_llm
#
###########################

# TODO : your code bellow

prompt_me = PromptTemplate.from_template("Please find the Date of Birth of {name}") #same as 2.1.2

llm_Ch2 = LLMChain(prompt=prompt_me,llm = davinci)

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Use `LLMChain.run()` or `LLMChain.predict()` to ask for the date of birth of Napoleon.


</div>

In [ ]:
###########################
# Task: 
#   Asks the language model to give the date of birth of a historical figure.
#
###########################

# TODO : your code bellow
llm_Ch2.run('Napolean')
llm_Ch2.predict(name = 'Su Shi')

'\n\nSu Shi, also known as Su Dongpo, was born 10 June 1037 in Meishan, Sichuan province, China.'

**SPOILER:** Alright that's better, Napoleon was indeed born of August 15, 1769. Let's ask something more complex and recent and see how our LLM does.

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Using [langchain.PromptTemplate](https://api.python.langchain.com/en/latest/prompts/langchain.prompts.prompt.PromptTemplate.html), create a new prompt template with an input variable called `year` that asks the language model to give the winner of the FIFA world cup on a given year.

</div>

In [ ]:
###########################
# Task: 
#   Create a PromptTemplate with an input variable called `year` that asks the language model to give the winner of the FIFA world cup of that year.
###########################

# TODO : your code bellow
prompt_me2 = PromptTemplate.from_template("Please find the Winner of the FIFA world cup of the year {year}") #same as 2.1.2


<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Using langchain.LLMChain, create a chain to run the new prompt template by our OpenAI LLM.

</div>

In [ ]:
###########################
# Task: 
#   Create a LLMchain with your prompt and your hub_llm
#
###########################

llm_Ch3 = LLMChain(prompt=prompt_me2,llm = davinci)


<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Use `LLMChain.run()` or `LLMChain.predict()` to ask for the winner of the 2022 FIFA world cup.


</div>

In [ ]:
###########################
# Task: 
#   Asks the language model to ask for the winner of the 2022 FIFA world cup
#
###########################

# TODO : your code bellow
llm_Ch3.run(2022)

'\n\nThe FIFA World Cup of 2022 has not yet taken place, so no winner has been determined.'

The LLM doesn't know what happened in 2022 because it was trained before that year. How can we update the LLM's knowledge to make it up to date? With Retrieval augmentation!

# 3. Retrieval Augmentation of a LLM using a Vector Database

**WARNING:** Some of the cells in this part may take some time to run as we are working with a lot of data.

The most powerful LLMs in the world have no idea about recent world events, nor can they cite sources. In general a LLM will only have knowledge about what it has been exposed to during training. For LLMs, the world exists as a static snapshot of the world as it was within their training data. 

A solution to this problem is **retrieval augmentation**: we retrieve relevant information from an external knowledge base and give that information to our LLM.

*Source: https://docs.pinecone.io/docs/langchain#retrieval-augmentation-in-langchain*


<div width=50% style="display: block; margin: auto">
    <img src="figures/augmentation.png" width=70%>
</div>

To perform **retrieval augmentation**, we will embed data with an **embedding model**, store the embedded data into a **vector database index**, and create a **LangChain vectorstore** that will use the index and the embedding model to find relevant information to the prompt sent by the user, and feed the most relevant results found in the database to the **LLM** to provide it with context and sources.

## 3.1 Building a knowledge base with Vector Embedding 

Source: [Creating the knowledge base](https://www.pinecone.io/learn/series/langchain/langchain-retrieval-augmentation/#Creating-the-Knowledge-Base)

This part concists in taking a dataset of relevant content we want to augment our LLM with (it could be code documentation for an LLM that needs to help write code, company documents for an internal chatbot...), process it, and embedded it into vectors. 

You can look [here](https://www.pinecone.io/learn/series/langchain/langchain-retrieval-augmentation/#Creating-the-Knowledge-Base) fore the full processs, which you would have to do if you wanted to augment your LLM for a specific use-case.

**For the sake of simplicity, we will use a pre-embembeded dataset from pinecone_dataset and upload it to a Pinecone Vector database. Unfortunatly, this dataset doesn't contain data on the 2022 FIFA world cup, but the process used here can be applied with other datasets, giving you an idea of the process should you want to use it for personal projects.**

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the cell bellow to import and process the pre-embedded data we will use for retrieval augmentation

</div>

In [60]:
# Run this cell
%pip install --user apache_beam



  Obtaining dependency information for apache_beam from https://files.pythonhosted.org/packages/43/c3/c35f521f79698d749c016179acf0718c6d9c8a6ecc9d2b45f84ff92c6be2/apache_beam-2.51.0-cp311-cp311-win_amd64.whl.metadata
  Using cached apache_beam-2.51.0-cp311-cp311-win_amd64.whl.metadata (6.1 kB)
  Using cached dill-0.3.1.1-py3-none-any.whl
  Using cached cloudpickle-2.2.1-py3-none-any.whl (25 kB)
  Obtaining dependency information for fastavro<2,>=0.23.6 from https://files.pythonhosted.org/packages/1c/d5/d8a47ac699fcdd83c2d54d095183a715170ce56162d67e451c6d61bf7565/fastavro-1.9.0-cp311-cp311-win_amd64.whl.metadata
  Using cached fastavro-1.9.0-cp311-cp311-win_amd64.whl.metadata (5.7 kB)
  Obtaining dependency information for fasteners<1.0,>=0.3 from https://files.pythonhosted.org/packages/61/bf/fd60001b3abc5222d8eaa4a204cd8c0ae78e75adc688f33ce4bf25b7fafa/fasteners-0.19-py3-none-any.whl.metadata
  Using cached fasteners-0.19-py3-none-any.whl.metadata (4.9 kB)
  Using cached hdfs-2.7.3-py3

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
multiprocess 0.70.15 requires dill>=0.3.7, but you have dill 0.3.1.1 which is incompatible.


In [3]:
#import pinecone_datasets

# We import a dataset
#dataset = pinecone_datasets.load_dataset('wikipedia-simple-text-embedding-ada-002-100K')

from datasets import load_dataset

dataset =  load_dataset("wikipedia", "20220301.simple", split='train[:10000]')

# We drop sparse_values as they are not needed for this example
#dataset.documents.drop(['metadata'], axis=1, inplace=True)
#dataset.documents.rename(columns={'blob': 'metadata'}, inplace=True)
# We will use rows of the dataset up to index 30_000 to make the upload to the Pinecone Vector Database faster
#dataset.documents.drop(dataset.documents.index[30_000:], inplace=True)

dataset

Dataset({
    features: ['id', 'url', 'title', 'text'],
    num_rows: 10000
})

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the cell bellow to check that your pinecone database does not contain any indexes, and delete them if there are any.

</div>

In [9]:


pinecone.list_indexes()

#pinecone.delete_index("index_name")

[]

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Use the [`create_index`](https://docs.pinecone.io/reference/create_index) function to create an index, name it `langchain-retrieval-augmentation-fast`, set the dimension to **1536**, and use the metric `cosine`.

</div>


**Note:** We set the dimension to 1536 as this is the dimension of OpenAI's text embedding model 'text-embedding-ada-002' that we use to embed the data. If you wish to use another model (for instance a Hugging Face model using HuggingFaceInferenceAPIEmbeddings), you will have to change this number to match the dimension of your model.

In [20]:
###########################
# Task: 
#   Check that your pinecone database does not contain any indexes, and delete them if there are any.
#
###########################
#pinecone.delete_index(name = index_name)

In [23]:

index_name = 'langchain-retrieval-augmentation-fast'

# TODO : your code bellow
try:
    pinecone.describe_index(name=index_name)
except:
    pinecone.create_index(name=index_name,dimension = 1536, metric='cosine')

pinecone.list_indexes()

pinecone.describe_index(name=index_name)



IndexDescription(name='langchain-retrieval-augmentation-fast', metric='cosine', replicas=1, dimension=1536.0, shards=1, pods=1, pod_type='starter', status={'ready': True, 'state': 'Ready'}, metadata_config=None, source_collection='')

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the cell bellow to push the data on to the Vector Database. This can take a few minutes.

</div>

In [41]:
# Run this cell - It may take a few minutes

import time

index_me = pinecone.GRPCIndex(index_name) # GRPC allows for faster upserts to the Pinecone Vector Database
# wait a moment for the index to be fully initialized
time.sleep(1)

index_me.describe_index_stats()

for batch in dataset.iter_documents(batch_size=100):
    index_me.upsert(batch)


index_me.describe_index_stats()

PineconeException: UNKNOWN:DNS resolution failed for langchain-retrieval-augmentation-fast-too68lo.svc.gcp-starter.pinecone.io:443: UNAVAILABLE: WSA Error {grpc_status:14, created_time:"2023-11-16T15:24:53.9107205+00:00"}

## 3.2 Creating a vector store

Now that we've build our index we can switch over to LangChain. We need to initialize a LangChain vector store using the same index we just built. For this we will also need a LangChain embedding object.

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the following cells to initialize a LangChain vector store.

</div>

In [37]:
# Run this cell

from langchain.embeddings.openai import OpenAIEmbeddings

model_name = 'text-embedding-ada-002'

embed = OpenAIEmbeddings(
    model=model_name,
    openai_api_key=OPENAI_API_KEY
)

In [26]:
# Run this cell

from langchain.vectorstores import Pinecone

text_field = "text"

# switch back to normal index for langchain
index_me = pinecone.Index(index_name)

vectorstore = Pinecone(
    index_me, embed, text_field
)



<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the following cell to perform a `similarity_search` of the content of the query on our [vectorstore](https://python.langchain.com/docs/modules/data_connection/vectorstores/) containing embedded information about our augmentation dataset.

</div>

In [27]:
# Run this cell

query = "When was Napoleon born"

vectorstore.similarity_search(
    query,  # our search query
    k=3  # return 3 most relevant docs
)

[]

## 3.3 Augmented LLM query

We finally have all the pieces of the puzzle, and we can now force the LLM to answer a question based on the information it is seeing being returned from the vectorstore. This allows for many things, like giving it up-to-date information, but can also be used to [cite sources](https://python.langchain.com/docs/use_cases/question_answering/vector_db_qa#return-source-documents). 

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Run the following cell to create a `VectorStoreRetrieverMemory` object that we will pass on to the LLMChain.
</div>

In [28]:
# Run this cell
from langchain.memory.vectorstore import VectorStoreRetrieverMemory

retriever = vectorstore.as_retriever(search_kwargs=dict(k=1))
memory_RAG = VectorStoreRetrieverMemory(retriever=retriever)


<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

- Create a new prompt template with a question of your choice. For example, you can ask about the life of historical figures.

- Using langchain.LLMChain, create a chain to run the new prompt template by our OpenAI LLM davinci. Add to the lists of arguments `memory = memory_RAG` to instruct the LLM to look for the answer in the vectorstore.



</div>

In [38]:
###########################
# Task: 
#   Asks the language model to give the date of birth of a historical figure.
#
###########################

# TODO : your code bellow
prompt_me = PromptTemplate.from_template("Please find the Date of Birth of {name}") #same as 2.1.2

llm_Ch2 = LLMChain(prompt=prompt_me,llm = davinci,memory = memory_RAG)

## 3.4 Augmented LLM query with sources

As stated before, we can use retrieval augmentation to provide sources to the output of the LLM (provided the sources were in the metadata of the embedded dataset).

In [40]:
# Run this cell

from langchain.chains import RetrievalQAWithSourcesChain

chain = RetrievalQAWithSourcesChain.from_chain_type(llm = davinci, chain_type="stuff", retriever=retriever)
chain({"question": "Tell me about the life of Napoleon"})


{'question': 'Tell me about the life of Napoleon',
 'answer': " I don't know.\n",
 'sources': ''}



<div class="alert alert-block alert-warning">
    <b>👩‍💻👨‍💻 Optional action</b>

- Reading: [Making Retrieval Augmented Generation Fast](https://www.pinecone.io/learn/fast-retrieval-augmented-generation/)
    
</div>

# 4. The end

That's the end of this lab! We hope you learned a lot through it, and that you are now ready to go on the adventure on your own and explore all that is possible to do with LLMs and vector Databases.

<div class="alert alert-block alert-warning">
    <b>👩‍💻👨‍💻 Optional action</b>

Explore the other exemple of applications of vector databases listed in Pinecone's documentation:

https://docs.pinecone.io/page/examples
    
</div>


To submit this assignment and **every other future assignment**, included the **final assignment** you have to:
- Commit and push your code to GitHub
- Go to **your** repository of the assignment. This must be on our course organisation `UCL-ELEC0136` and usually has the pattern `https://github.com/UCL-ELEC0136/<assignment-name>-<your-github-username>`.
- Go in the `Pull requests` tab and click on the `Feedback` pull request.
- Click on `Files changed` and verify that the files you have changed are listed.
- Merge the pull request by clicking on `Merge pull request` and then `Confirm merge`.

We are now ready to push our code that acquires data from GitHub to our repository (which is also GitHub, but this is just a coincidence, we could have used any other API, like Twitter's or Facebook's).

<div class="alert alert-block alert-danger">
<b>👩‍💻👨‍💻 Action required</b>

Submit your assignment by following the steps above.
</div>